### ISSUE TITLE
Kafka key is not populated when using kafka-binding with Flux

### ISSUE THREAD

**Issue by sabbyanandan** [spring-cloud/spring-cloud-stream-binder-kafka#972]
@rgonciarz commented on [Thu Oct 08 2020](https://github.com/spring-cloud/spring-cloud-stream/issues/2030)

**Describe the issue**
Kafka key is not populated when using kafka-binding with Flux

```
    @Bean
    fun topologyR() = Function<Flux<String>,Flux<String>> {
        it.doOnNext { v -> log.info("topologyR: {}", v) }
    }
```

Setup:
topic1 -> topologyA (kafka streams binding) -> topic2 
topic2 -> topologyR (kafka flux binding) -> topic3
topic3 -> topologyB (kafka streams binding)

**To Reproduce**
I created a small project that may help to reproduce the issue:
https://github.com/rgonciarz/reactive-topology-chain/tree/reactive-function-no-avro](https://github.com/rgonciarz/reactive-topology-chain/tree/reactive-function-no-avro)

1. Run application, and either send a message on topic1 (String,String serialization) or use ApplicationKtTest 

Result:
```
2020-10-08 19:20:02.221  INFO 14232 --- [-StreamThread-1] wr.reactive.TopologyConfig               : topologyA: 9307f81e-7148-459e-a49a-e5aeb07cacac, 9307f81e-7148-459e-a49a-e5aeb07cacac
2020-10-08 19:20:02.324  INFO 14232 --- [container-0-C-1] wr.reactive.TopologyConfig               : topologyR: 9307f81e-7148-459e-a49a-e5aeb07cacac
2020-10-08 19:20:02.326  INFO 14232 --- [-StreamThread-1] wr.reactive.TopologyConfig               : topologyB: null, 9307f81e-7148-459e-a49a-e5aeb07cacac
```
We loose information about Kafka key. To workaround the problem we need to use Message abstraction and manually copy header from KafkaHeaders.RECEIVED_MESSAGE_KEY to KafkaHeaders.MESSAGE_KEY.

Affected version: Spring boot 2.3.3

---

@garyrussell commented on [Mon Oct 12 2020](https://github.com/spring-cloud/spring-cloud-stream/issues/2030#issuecomment-707227465)

The framework cannot make any assumption that the inbound key should be propagated to the outbound key.

Perhaps an option could be added, but your work around is the correct way to handle this currently.

**Comment by garyrussell** [comment]
>Perhaps an option could be added, but your work around is the correct way to handle this currently.

However, this would only apply to, e.g. `Function<String, String>` (not `Flux`).

**Comment by garyrussell** [comment]
This should just need

    spring.cloud.stream.kafka.bindings.input-out-0.producer.messageKeyExpression=headers['kafka_receivedMessageKey']

but it doesn't work because the inbound headers are not copied to the outbound message.

https://github.com/spring-cloud/spring-cloud-function/issues/599

**Comment by sobychacko** [comment]
@rgonciarz This issue is now resolved on both 3.1.x (master) and 3.0.x branches based on changes in Spring Cloud Function. We used this issue to explain further the usage of `messageKeyExpression`. 

Basically, given a function as:

```
@Bean
public Function<Message<?>, Message<?>> topologyR() {
  return m -> m;
}
```
or
```
@Bean
public Function<String, String> topologyR() {
  return m -> m;
}
```

and the property: `spring.cloud.stream.kafka.bindings.topologyR-out-0.producer.messageKeyExpression: headers['kafka_receivedMessageKey']`, Spring Cloud Stream will produce the message with the same incoming key. 

This will not work for reactive functions though as it is up to the application to copy the headers manually using `Message<?>`.
